src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,224 @@
+   * Resolves when time is set.
+   * @param {number} time
+   * @returns {Promise}
+   */
+  setTime(time: number): Promise<unknown> {
+    for (let i = 0; i < this._videoElement.buffered.length; i++) {
+      if (this._videoElement.buffered.start(i) <= time &&
+          this._videoElement.buffered.end(i) >= time) {
+        this._videoElement.currentTime = time;
+        log.debug("VTL: Thumbnail already loaded.");
+        return PPromise.resolve(time);
+      }
+    }
+
+    const manifest = this._player.getManifest();
+    if (manifest === null) {
+      return PPromise.reject(new VideoThumbnailLoaderError("NO_MANIFEST",
+                                                           "No manifest available."));
+    }
+    const contentInfos = getContentInfos(time, manifest);
+    if (contentInfos === null) {
+      return PPromise.reject(new VideoThumbnailLoaderError("NO_TRACK",
+                                                           "Couldn't find track for this time."));
+    }
+    const initURL = contentInfos.representation.index.getInitSegment()?.mediaURLs?.[0] ?? "";
+    if (initURL === "") {
+      return PPromise.reject(new VideoThumbnailLoaderError("NO_INIT_DATA", "No init data for track."));
+    }
+    const segment = contentInfos.representation.index.getSegments(time, time + 10)[0];
+    if (segment === undefined) {
+      return PPromise.reject(new VideoThumbnailLoaderError("NO_THUMBNAILS", "Couldn't find thumbnail."));
+    }
+
+    log.debug("VTL: Found thumbnail for time", time, segment);
+
+    if (this._currentJob !== undefined &&
+        this._currentJob.contentInfos.representation.id ===
+          contentInfos.representation.id &&
+        this._currentJob.segment.time === segment.time &&
+        this._currentJob.segment.duration === segment.duration &&
+        this._currentJob.segment.mediaURLs?.[0] === segment.mediaURLs?.[0]) {
+      // The current job is already handling the loading for the wanted time (same
+      // thumbnail).
+      return this._currentJob.jobPromise;
+    }
+    this._currentJob?.stop();
+    return this.startJob(contentInfos, time, segment);
+  }
+
+  /**
+   * Dispose thumbnail loader.
+   * @returns {void}
+   */
+  dispose(): void {
+    this._currentJob?.stop();
+    disposeSourceBuffer(this._videoElement);
+  }
+
+  private startJob = (contentInfos: IContentInfos,
+                      time: number,
+                      segment: ISegment
+  ): Promise<unknown> => {
+    const killJob$ = new Subject();
+
+    const abortError$ = killJob$.pipe(
+      map(() => {
+        throw new VideoThumbnailLoaderError("ABORTED",
+                                            "VideoThumbnailLoaderError: Aborted job.");
+      })
+    );
+
+    const jobPromise = observableRace(
+      initSourceBuffer$(contentInfos,
+                        this._videoElement).pipe(
+        mergeMap((videoSourceBuffer) => {
+          return removeBuffer$(this._videoElement, videoSourceBuffer, time).pipe(
```

Comment:
To speed up the process, this can be made at the same time than the request.

Or maybe you want to guarantee the memory level to be low before doing a segment request?